### ISSUE TITLE
[Feature Request] Support for direct CLONING from inside lazygit

### ISSUE THREAD

**Issue by iocouto** [jesseduffield/lazygit#1257]
Right now we are already able to create a new local repository with lazygit, as well as manage a previously-cloned one. It seems odd that I'm not able to also **clone** a repository with lazygit...

**Comment by mark2185** [comment]
@jesseduffield this might be an option in the `Status` pane that brings up a prompt for the link and destination path?

Seems like a good first issue.

**Comment by jesseduffield** [comment]
Agreed

**Comment by jesseduffield** [comment]
we really need to introduce a form panel

**Comment by willparsons** [comment]
should this be delayed until the form panel has been created?

**Comment by willparsons** [comment]
once the repo has been cloned do we want to
 - do nothing
 - ask the user if they want to switch to the cloned repo
 - put the repo into the "recent repos" list

**Comment by jesseduffield** [comment]
@willparsons it's okay for us to have basic git clone functionality. We can add the form approach later. After cloning we should ask the user if they want to switch to the cloned repo. We should also (unconditionally) add the repo to the recent repos list

**Comment by JordanllHarper** [comment]
I'm interested in picking this up! Just want to be clear on my assumptions going into this (still relatively new to open source and the repo 😄 ):

- Invoking this would be on the status pane, using a hotkey like `c`.
- We'd get the default prompt for putting in a url.
- Hitting enter on that would clone the repo to the current directory and then a popup would appear asking to switch - enter to switch, esc to not switch.
- We also add this to the recent repos list, no matter if the user switches or not.

**Comment by stefanhaller** [comment]
> * Hitting enter on that would clone the repo to the current directory

This doesn't work. The current directory is always the one of the currently open repo, you don't want to clone into that.

If anything, you _could_ remember the path where lazygit was started, and use that. This still doesn't work if lazygit was started from within an existing repo, which for me personally is the case 100% of the time. You'd then have to show some kind of directory chooser, which we don't have, or just a prompt where people can type in the destination path (this was mentioned above).

But honestly, why? It's so much easier to open a new terminal prompt, cd into the directory you want to clone into (and there's tab completion in your shell to make this easy, something that we don't want to build into lazygit), type `git clone <paste>` if the URL is in your clipboard, cd into the cloned directory, and start lazygit there.

I just don't see why it's worth spending the time on building this into lazygit, when it's more convenient in the shell.

**Comment by stefanhaller** [comment]
Other ideas:
- use the parent of the currently open directory as the destination, assuming that's where you keep all your repos (won't work well if the currently open repo is a submodule of another one)
- take the parents of all the repos in your recent repos list, remove duplicates (and remove the ones that point inside other repos in the list, to take care of the submodule issue), and present the resulting list in a menu to pick from.

**Comment by JordanllHarper** [comment]
You definitely bring up some good points, though I like the idea of this happening if you're not in a Git repository already. 

In my head, a typical flow you might go through:

You open the terminal, open lazygit, get the terminal prompt with an ability to clone. Then paste a url, hit enter and you immedately change working directory into the newly cloned project and open the lazygit interface to make branches or whatever you want to do.

Of course, you could solve this with some custom aliases, but this might be a nice option.

**Comment by christianschult** [comment]
Since it seems like this isn't being worked on at the moment, I'd like to pick up this issue, provided this is still something you could see being added to the project.
I prefer the idea of this being an option when opening lazygit outside of a repo as it makes more sense to me than cloning a repo while already working on a different one. Lazygit has several config options for how to behave when being opened outside of a repo with "prompt", "create", "skip" and "quit". An additional option like "clone" doesn't seem out of place here. I also think that it would be nice to have this included in lazygit since it already handles all the other git-related functionality.

**Comment by stefanhaller** [comment]
Adding "clone" as an option for the "notARepository" config doesn't make sense to me. Users will probably want to decide case-by-case whether they want to open a recent repo, create a new one, or clone one. So if we wanted to add this functionality, we'd probably have to turn the current "Create a new git repository?" prompt into a multi-choice menu.

I still don't see the value of this feature though when it is so easy to type "git clone <paste>" in the terminal, so honestly I'd prefer not to spend more time designing this right now.

**Comment by Duckelekuuk** [comment]
The irony of saying that typing ``git clone <repo>`` is easy while the project is a UI for those commands. I see the benefit of adding this feature. I found this issue because I was looking for the feature myself.

**Comment by JordanllHarper** [comment]
Agreed, by the same logic it's easy to init a new repository but we have that built into the app too. 

Sure there's some complexity in regards to picking which option but isn't this the point of lazygit? To make this stuff easy and fast?

**Comment by stefanhaller** [comment]
> The irony of saying typing `git clone <repo>` is easy while the project is about creating a UI wrapper for the git CLI.

You have a point here. However, adding the feature to lazygit only makes sense if we can do it in such a way that it is actually easier (more straightforward, easier to understand, less typing) than using the command line, and so far I haven't seen a suggestion that ticks these boxes.

When I said "I don't see the value of this feature", what I meant is I don't object to this being added to lazygit; however, I don't see enough value in it to spend time on helping design the feature to make it really awesome, and I do object to adding it with a half-baked design.

> Agreed, by the same logic it's easy to init a new repository but we have that built into the app too.

Yes, and I don't like it much. 😄  (This was added way before I joined the project.)
